src/v/utils/vint.h

```diff
@@ -12,38 +12,73 @@
 #pragma once
 #include "bytes/bytes.h"
 
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
 #include <cstdint>
 
 namespace unsigned_vint {
 /// At most 5 bytes are needed to encode a 32 bit value
 inline constexpr size_t max_length = 5;
 
 namespace detail {
-
-/// \brief almost identical impl to leveldb, made generic for c++
-/// friendliness
-/// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
-template<typename Range>
-inline std::pair<uint64_t, size_t>
-deserialize(Range&& r, uint8_t limit) noexcept {
+struct var_decoder {
     uint64_t result = 0;
     size_t bytes_read = 0;
     uint64_t shift = 0;
-    for (auto src = r.begin(); shift <= limit && src != r.end(); ++src) {
-        uint64_t byte = *src;
+    uint8_t limit;
+
+    explicit var_decoder(uint8_t limit)
+      : limit(limit) {}
+
+    /**
+     * @brief Incorporate a new byte into the state.
+     * Returns true if the outer loop should break (last byte found).
+     */
+    bool accept(uint8_t byte) {
+        /// \brief almost identical impl to leveldb, made generic for c++
+        /// friendliness
+        /// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
         bytes_read++;
         if (byte & 128) {
             result |= ((byte & 127) << shift);
         } else {
             result |= byte << shift;
-            break;
+            return true;
         }
         shift += 7;
+        return shift > limit;
+    }
+};
+
+template<typename Range>
+inline std::pair<uint64_t, size_t>
+deserialize(Range&& r, uint8_t limit) noexcept {
+    var_decoder decoder(limit);
+    for (auto byte : r) {
+        if (decoder.accept(static_cast<uint8_t>(byte))) {
+            break;
+        }
     }
-    return {result, bytes_read};
+    return std::make_pair(decoder.result, decoder.bytes_read);
 }
+
 } // namespace detail
 
+inline ss::future<std::pair<uint32_t, size_t>>
+deserialize(ss::input_stream<char>& s) {
+    detail::var_decoder decoder(5 * sizeof(uint64_t));
```

Comment:
Where does this constant `40` come from (`5 * sizeof(uint64_t)`)

Reply:
Nice catch, that should be [4 * 5](https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L89). Is it because at most 5 iterations of reading a `uint32_t` would be the maximum number of iterations to perform when deserializing a uint

Reply:
Thanks. Sounds like it needs a unit test? Trying to follow the math, each byte consumed encodes 7 bits of the varint, to get a `u32` you'd need `5` bytes (`7 * 5 = 35 < 32`).. am I missing something?

Reply:
Sorry yes I linked to a number that == `28` yet I said `4*5` haha, I was needing coffee when I wrote that. So the number `28` pretty much is a stop gap saying that at most `28` bits should be read by this algorithm. It however does this by reading 1 `uint_8` 4 times for which you would have read read `32` bits at that point, but 4 of those will count as continuation bits. So 32 - 4 = 28. 

Good idea will add test